### ISSUE TITLE
Describe how a client may abandon a transfer if it cannot resume it

### ISSUE THREAD

**Issue by qsorix** [tus/tus-resumable-upload-protocol#52]
Today at work we've found a corner case when the client knows it was uploading something, but is unable to resume that.

Should TUS specify a way to gracefully abort an upload session? Of course it cannot be enforced, but in case of a closed environment, you may want to implement it, to release resources on the server sooner.

I'm fine if we decide it should be out of scope. But I guess it can be as simple as sending a request using `DELETE` instead of `PATCH`, so perhaps we can specify that? We only need to think how to respond to further PATCH and/or GET requests, but I guess `404` is the answer here.

Opinions?

**Comment by Acconut** [comment]
Good catch.

One solution would be to wait until the server cleans up expired and unfinished uploads. The bad part is that it only works if the server implements this behaviour (see upload-expiration extension).

A `DELETE` request to the file upload URL should remove the upload and free used resources. Any further request to the file URL should result in a `404 Not Found` or `410 Gone` if the server keeps track of deleted uploads.

I'm not sure if this should be in the core or defined as an extension. What do you prefer?

**Comment by qsorix** [comment]
> I'm not sure if this should be in the core or defined as an extension. What do you prefer?

I'm in control of all clients and servers that will ever communicate with each other in my environment. I'm in a position where we'll decide the client's are going to do this and then we'll have to implement it, or we'll work around or ignore client's limitations for the sake of simplicity and then there'll be no need to implement it at all. Whether it is in core or extension doesn't make a difference to me, so I may not be the right person to answer that question.

Do we have a rule about that's core and what's an extension? Let's say that if the goal of TUS is to define a resumable upload, then anything not essential to that task is an extension. This `DELETE` thing would then be an extension as well.

On the other hand, having too many trivial extension can make the protocol seem complicated.

So far, I think, the idea is to have more extensions than a big core. I don't mind. Let's go in that direction for 1.0 We can always change our mind and integrate some extensions into core later.

**Comment by Acconut** [comment]
I think the best way to go for this is as an extensions. The reason is that not every server needs this feature (so it should be optional) and this way we can take advantage of the `TUS-Extension` header for discovery supported extension.

Would you be so kind and propose changes required for this extension?

**Comment by qsorix** [comment]
> Would you be so kind and propose changes required for this extension?

With pleasure. But I'm going hiking for the weekend so I'm going to work on this only after Monday.

**Comment by Acconut** [comment]
Great, I will give it a shot over the weekend, too. Have a great time.

**Comment by Acconut** [comment]
# Abortion

This extensions defines a way for clients to abort unfinished uploads which won't be continued allowing servers to free up used resources. When receiving a `DELETE` request the server MUST reject all future uploads returning `404 No Found` or `410 Gone`. If this extension is supported by the server it MUST be announced by adding the `abortion` element to the `TUS-Extension` header.

-----

What do you think?

**Comment by qsorix** [comment]
Thanks for working on this. It's cool the way it is. I had a few comments, but it turned out it was more readable if I just edited the text. So based on the above, I ended up with:

# Abortion

This extensions defines a way for clients to abort unfinished uploads which
won't be continued allowing servers to free up used resources. All clients are
encouraged to implement this.

Clients can abort an upload by sending a `DELETE` request to the upload's url.

When the server receives a `DELETE` request for an existing upload, it CAN free
associated resources an MUST return `2xx` status code, confirming that the upload
was aborted. For all future requests to that URL the server MUST return `404 No Found` or `410
Gone`.

If this extension is supported by the server it MUST be announced by adding the
`abortion` element to the `TUS-Extension` header.

## Example 
**Request:**

```
DELETE /files/24e533e02ec3bc40c387f1a0e460e216 HTTP/1.1
Host: tus.example.org
Content-Length: 0
```

**Response:**

```
HTTP/1.1 204 No Content
```

**Comment by qsorix** [comment]
> If this extension is supported by the server it MUST be announced by adding the
abortion element to the TUS-Extension header.

IMHO if server doesn't announce it, and clients sends DELETE, it is still OK to remove the URL. I would perhaps change "MUST" to "SHOULD".

**Comment by Acconut** [comment]
I like your changes in general although it's mostly the type of language. Anyway I have some (not really important) comments:

> it CAN free associated resources

I think that's obvious. But if we add it why not use "SHOULD" in this case?

> an[d] MUST return 2xx status code, confirming that the upload was aborted

Maybe we should add that it must be a 204 (or 200).

The example should contain the `TUS-Resumable` header since #48 is merged.

> IMHO if server doesn't announce it, and clients sends DELETE, it is still OK to remove the URL. I would perhaps change "MUST" to "SHOULD".

Frankly, I don't understand your argument. Announcing support for abortion is required if you want to know whether the server this extension in order to display the according UI. Of course, the client can test support by sending a `DELETE` request but that's why we have introduced the `TUS-Extension` header.

**Comment by qsorix** [comment]
Hm, right. In short, I agree with all your comments. :)

**Comment by Acconut** [comment]
See #54.

**Comment by Acconut** [comment]
Merged.
